client/my-sites/stats/activity-log-tasklist/plugins-to-update.jsx

```diff
@@ -0,0 +1,56 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import { unionBy } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { getPluginsWithUpdates } from 'state/plugins/installed/selectors';
+
+export default WrappedComponent => {
+	class PluginsToUpdate extends Component {
+		static propTypes = {
+			siteId: PropTypes.number,
+			pluginsWithUpdates: PropTypes.arrayOf( PropTypes.object ),
+		};
+
+		state = {
+			// Plugins with pending updates
+			pluginsWithUpdates: [],
+			// Plugins already updated + those with pending updates
+			allPluginsToUpdate: [],
+		};
+
+		componentWillReceiveProps( nextProps ) {
+			if ( this.state.pluginsWithUpdates.length < nextProps.pluginsWithUpdates.length ) {
+				this.setState( {
+					pluginsWithUpdates: nextProps.pluginsWithUpdates,
+					allPluginsToUpdate: unionBy(
+						nextProps.pluginsWithUpdates,
+						this.state.allPluginsToUpdate,
+						'slug'
+					),
+				} );
+			}
+		}
```

Comment:
note that this won't run on mount. it would be good to use the new `getDerivedStateFromProps()` function here so that it _will_ and also because `componentWillReceiveProps()` is now officially being deprecated.

```js
getDerivedStateFromProps( nextProps, prevState ) {
	return {
		pluginsWithUpdates: nextProps.pluginsWithUpdates,
		allPluginsToUpdate: unionBy( nextProps.pluginsWithUpdates, prevState.allPluginsToUpdate, 'slug' )
	};
}
```

I'd also like to encourage you to take advantage of the fact that we have multiple levels of namespacing going on here already and that long names aren't always helpful. terse names can make it easy for many people to quickly ascertain the purpose a variable or function serves. in this case, we already know we're dealing with plugins to update because that's the file name and that's also the component name. Something terser can be easier to digest…

```js
getDerivedStateFromProps( nextProps, prevState ) {
	return {
		pending: nextProps.plugins,
		plugins: unionBy( nextProps.plugins, prevState.plugins, 'slug' )
	};
}

return <Wrapped siteId={ siteId } updatablePlugins={ this.state.pending } plugins={ this.state.plugins } />
```

finally, I'm confused on the name `allPluginsToUpdate` because our comments indicate that it involves not only plugins to update but also plugins which _were_ updated and no longer have any updates available. any way we could refine that name to be more specific or more general so that the contradiction disappears?

Reply:
Cool. I actually tried using `getDerivedStateFromProps` a couple of weeks ago and it wasn't available. I see that react and react-dom are now at 16.3.1 so will update.
Will review the names and about `allPluginsToUpdate`, I have nothing. `allPlugins` is not good because they're not all plugins, just those that will or have been updated now. Full name would be `allPluginsUpdatedOrToUpdate` and it's quite long. I'm open to suggestions since I also think it's not a good enough name.

Reply:
Remember that we have the file and component as a namespace, so we don't need to be super redundant inside with our names. It's okay to say that something like `allPlugins` or even `plugins` by itself has a specific meaning inside our component/domain.

`allPluginsUpdatedOrToUpdate` is long and redundant and can distract our train of thought when reading code: this why being so precise is not always helpful.

Reply:
This is now fixed and I've refactored things to get rid of the prop for pending plugin updates. Now there's a single one `plugins` that has the complete list of plugins with pending updates and those that were updated now.